platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+			 param->model, param->size);
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	if (odp_unlikely(param->max_compl_id > ML_MAX_COMPL_ID)) {
+		_ODP_ERR("param->max_compl_id: %u exceeds maximum completion id supported: %d\n",
+			 param->max_compl_id, ML_MAX_COMPL_ID);
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	/* Find an emtpy slot to store the new model */
+	for (i = 0; i < ML_MAX_MODELS_CREATED; i++) {
+		if (_odp_ml_glb->models[i].state)
+			continue;
+
+		odp_ticketlock_lock(&_odp_ml_glb->models[i].lock);
+
+		if (_odp_ml_glb->models[i].state) {
+			odp_ticketlock_unlock(&_odp_ml_glb->models[i].lock);
+			continue;
+		}
+
+		mdl = &_odp_ml_glb->models[i];
+		break;
+	}
+
+	if (i == ML_MAX_MODELS_CREATED) {
+		_ODP_ERR("Maximum number of models has already been created!\n");
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	/* Free model entry was found and is now locked */
+	mdl->state = ML_STATE_CREATED;
+
+	status = ort_api->CreateSessionOptions(&session_opts);
+	if (check_ortstatus(status) || !session_opts) {
+		_ODP_ERR("Error: CreateSessionOptions failed.\n");
+		mdl->state = ML_STATE_FREE;
+		odp_ticketlock_unlock(&mdl->lock);
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	if (set_ort_run_opts(name, session_opts)) {
+		_odp_ml_glb->ort_api->ReleaseSessionOptions(session_opts);
+		mdl->state = ML_STATE_FREE;
+		odp_ticketlock_unlock(&mdl->lock);
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	/* Store model info */
+	info = &mdl->info;
+	memset(info, 0, sizeof(odp_ml_model_info_t));
+
+	if (create_ort_model(param, &session, mdl, session_opts)) {
+		mdl->state = ML_STATE_FREE;
+
+		/* Initialize info back to 0 when some fields have been filled
+		 * while later failed */
+		reset_mdl_info_sizes(mdl);
+		odp_ticketlock_unlock(&mdl->lock);
+
+		_odp_ml_glb->ort_api->ReleaseSessionOptions(session_opts);
+		_ODP_ERR("create_ort_model() failed\n");
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	if (check_io_shape(mdl)) {
+		mdl->state = ML_STATE_FREE;
+		reset_mdl_info_sizes(mdl);
+		odp_ticketlock_unlock(&mdl->lock);
+
+		ort_api->ReleaseSession(session);
+		_odp_ml_glb->ort_api->ReleaseSessionOptions(session_opts);
+		return ODP_ML_MODEL_INVALID;
+	}
+
+	mdl->session = session;
+	mdl->session_opts = session_opts;
+	info->index = i;
+
+	if (name) {
+		strncpy(info->name, name, ODP_ML_MODEL_NAME_LEN - 1);
+		info->name[ODP_ML_MODEL_NAME_LEN - 1] = 0;
+	}
+
+	mdl->max_compl_id = param->max_compl_id;
+	for (uint32_t j = 0; j < ML_MAX_COMPL_ID; j++)
+		odp_atomic_init_u32(&mdl->compl_status[j], 1);
+
+	odp_ticketlock_unlock(&mdl->lock);
+	return (odp_ml_model_t)mdl;
+}
+
+int odp_ml_model_destroy(odp_ml_model_t model)
+{
+	ml_model_t *mdl = ml_model_from_handle(model);
+
+	if (model == ODP_ML_MODEL_INVALID) {
+		_ODP_ERR("Bad ML model handle\n");
+		return -1;
+	}
+
+	odp_ticketlock_lock(&mdl->lock);
+
+	if (mdl->state == ML_STATE_FREE) {
+		_ODP_ERR("Model not created\n");
+		odp_ticketlock_unlock(&mdl->lock);
+		return -1;
+	}
+
+	if (mdl->state == ML_STATE_LOADED) {
```

Comment:
Only acceptable state is "created", maybe simply check against that. Currently, check misses inferencing mode, which also a "loaded" state.